drivers/adc/mcp356x/mcp356x.c

```diff
@@ -0,0 +1,545 @@
+    }
+    printk("%s: " BYTE_TO_BINARY_PATTERN "\n", names[reg],
+           BYTE_TO_BINARY(data));
+  }
+
+  uint32_t timer = 0;
+  mcp_read_reg_24(config, MCP_Reg_TIMER, &timer);
+  printk("TIMER: %d\n", timer);
+
+  return 0;
+}
+
+static int mcp356x_init(const struct device *dev) {
+  const struct mcp356x_config *config = dev->config;
+  struct mcp356x_data *data = dev->data;
+  data->enabled_channels_bitmap = 0;
+
+  if (!device_is_ready(config->bus.bus)) {
+    LOG_ERR("SPI bus '%s'not ready", config->bus.bus->name);
+    return -ENODEV;
+  }
+
+  // Page 91 CONFIG0 ----------------------------------------------------------
+  // VREF Selection done by channel but 1 is the default of the adc
+  uint8_t vref_sel_bits = 0b1;
+
+  // CLK Selection
+  if (config->clock > 3) {
+    LOG_ERR("Unknown clock selection bits %d", config->clock);
+    return -ENOTSUP;
+  }
+  uint8_t clk_sel_bits = (uint8_t)config->clock;
+
+  // No Current Source/Sink Selection Bits for Sensor Bias
+  uint8_t adc_mode = 0b11; // continous conversion mode
+  data->config0 =
+      (vref_sel_bits << 7) | (1 << 6) | (clk_sel_bits << 4) | adc_mode;
+  mcp_write_reg_8(config, MCP_Reg_CONFIG0, data->config0);
+
+  // Page 92 CONFIG1 ----------------------------------------------------------
+  // Configure OSR
+  uint8_t osr_bits = (uint8_t)config->osr;
+
+  // Configure PRESCALE
+  if (config->prescale > 3) {
+    LOG_ERR("Invalid prescale selection %d", config->prescale);
+    return -ENOTSUP;
+  }
+  uint8_t prescale_bits = (uint8_t)config->prescale;
+
+  data->config1 = (prescale_bits << 6) | (osr_bits << 2);
+
+  mcp_write_reg_8(config, MCP_Reg_CONFIG1, data->config1);
+
+  // Page 93 CONFIG2 ----------------------------------------------------------
+  // Configure Gain
+  uint8_t gain_bits = 0;
+  uint8_t boost = 0b10; // 1x boost (bias current) (default)
+  uint8_t az_mux = 0b0; // mux auto zero internal (default)
+  uint8_t az_ref = 0b1; // auto zero internal voltage ref (default)
+  uint8_t reserved_bit = 0b1;
+
+  data->config2 = (boost << 6) | (gain_bits << 3) | (az_mux << 2) |
+                  (az_ref << 1) | reserved_bit;
+  mcp_write_reg_8(config, MCP_Reg_CONFIG2, data->config2);
+
+  // Page 94 CONFIG3 ----------------------------------------------------------
+  uint8_t conv_mode = 0b11;
+  uint8_t data_format = 0b00; // 24 bit adc data (default)
+  uint8_t crc_format = 0b0;   // 16 bit crc format (default)
+  uint8_t en_crccom = 0b0;    // offset calibration (default)
+  uint8_t en_offcal = 0b0;    // offset calibration (default)
+  uint8_t en_gaincal = 0b0;   // gain calibration (default)
+  data->config3 = (conv_mode << 6) | (data_format << 4) | (crc_format << 3) |
+                  (en_crccom << 2) | (en_offcal << 1) | en_gaincal;
+
+  (void)mcp_write_reg_8(config, MCP_Reg_CONFIG3, data->config3);
+
+  uint8_t irq_reg =
+      0b00110111; // set irq to active low so we don't need a pull up resistor
+  mcp_write_reg_8(config, MCP_Reg_IRQ, irq_reg);
```

Comment:
This probably gets optimized as text replacement at compile time but I would make `irq_reg` into a macro def and put it in `mcp_write_reg_8` directly. 

```c
#define IRQ_REG 0b00110111
...
mcp_write_reg_8(config, MCP_Reg_IRQ, IRQ_REG);

```